pl_bolts/models/rl/advantage_actor_critic_model.py

```diff
@@ -0,0 +1,327 @@
+        """
+        Calculates the loss for A2C which is a weighted sum of
+        actor loss (MSE), critic loss (PG), and entropy (for exploration)
+
+        Args:
+            states: tensor of shape (batch_size, state dimension)
+            actions: tensor of shape (batch_size, )
+            returns: tensor of shape (batch_size, )
+        """
+
+        logprobs, values = self.net(states)
+
+        # calculates (normalized) advantage
+        with torch.no_grad():
+            # critic is trained with normalized returns, so we need to scale the values here
+            advs = returns - values * returns.std() + returns.mean()
+            # normalize advantages to train actor
+            advs = (advs - advs.mean()) / (advs.std() + self.eps)
+            # normalize returns to train critic
+            targets = (returns - returns.mean()) / (returns.std() + self.eps)
+
+        # entropy loss
+        entropy = -logprobs.exp() * logprobs
+        entropy = self.hparams.entropy_beta * entropy.sum(1).mean()
+
+        # actor loss
+        logprobs = logprobs[range(self.hparams.batch_size), actions]
+        actor_loss = -(logprobs * advs).mean()
+
+        # critic loss
+        critic_loss = self.hparams.critic_beta * torch.square(targets - values).mean()
+
+        # total loss (weighted sum)
+        total_loss = actor_loss + critic_loss - entropy
+        return total_loss
+
+    def training_step(self, batch: Tuple[Tensor, Tensor], batch_idx: int) -> OrderedDict:
+        """
+        Perform one actor-critic update using a batch of data
+
+        Args:
+            batch: a batch of (states, actions, returns)
+        """
+        states, actions, returns = batch
+        loss = self.loss(states, actions, returns)
+
+        log = {
+            "episodes": self.done_episodes,
+            "reward": self.total_rewards[-1],
+            "avg_reward": self.avg_rewards,
+        }
+        return OrderedDict({
+            "loss": loss,
+            "avg_reward": self.avg_rewards,
+            "log": log,
+            "progress_bar": log,
+        })
+
+    def configure_optimizers(self) -> List[Optimizer]:
+        """Initialize Adam optimizer"""
+        optimizer = optim.Adam(self.net.parameters(), lr=self.hparams.lr)
+        return [optimizer]
+
+    def _dataloader(self) -> DataLoader:
+        """Initialize the Replay Buffer dataset used for retrieving experiences"""
+        dataset = ExperienceSourceDataset(self.train_batch)
+        dataloader = DataLoader(dataset=dataset, batch_size=self.hparams.batch_size)
+        return dataloader
+
+    def train_dataloader(self) -> DataLoader:
+        """Get train loader"""
+        return self._dataloader()
+
+    def get_device(self, batch) -> str:
+        """Retrieve device currently being used by minibatch"""
+        return batch[0][0][0].device.index if self.on_gpu else "cpu"
+
+    @staticmethod
+    def add_model_specific_args(arg_parser: ArgumentParser) -> ArgumentParser:
```

Comment:
Consider using `LightningCLI` see: https://devblog.pytorchlightning.ai/auto-structuring-deep-learning-projects-with-the-lightning-cli-9f40f1ef8b36

Reply:
Hi, I looked a bit into it, and It seems like in order to add the arguments (`add_model_specific_args()`) in the lightning module, I'll have to write my own subclass of `LightningCLI`, which defeats the purpose to use it to simplify the code. Let me know if my understanding is wrong. Or is there an example of using it in other RL algorithms?

Reply:
the `LightningCLI` takes and maps what ever you have in Module init...